### PR TITLE
prepare for mapdb backend

### PR DESCRIPTION
**Pull request by javeme**
Change-Id: I046bde4a32b3fd6693f788d27a9165e2646bc5cb

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java [hunk 1]
```diff
@@ -311,32 +311,40 @@ public void rollback() throws BackendException {
...
-            } else if (!q.empty()) {
-                // Return empty if there is no result after index-query
-                queries.add(q);
+            Query q = this.optimizeQuery(cq);
+            // Return empty if there is no result after index-query
+            if (q.empty()) {
+                return Collections.emptyIterator();
+            }
+            return super.query(q);
+        } else {
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java]: Remove the else, the code inside the brackets is slightly longer

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java [hunk 2]
```diff
@@ -604,7 +612,9 @@ public void removeEdge(HugeEdge edge) {
     public Iterator<Edge> queryEdges(Query query) {
         Iterator<HugeEdge> results = this.queryEdgesFromBackend(query);
 
-        Set<Id> returnedEdges = new HashSet<>();
+        // TODO: any unconsidered case?
+        boolean withDuplicatedEdge = false;
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java]: When will it be true?

Comment (javeme) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java]: maybe the query with OR condition, like query {OUT edges of vertex1} OR {IN edges of vertex2}

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BytesBuffer.java [hunk 3]
```diff
@@ -48,6 +48,10 @@
     public static final int ID_MAX_LEN = UINT8_MAX & 0x7f + 1; // 128
     public static final int BIG_ID_MAX_LEN = UINT16_MAX & 0x7fff + 1; // 32768
 
+    public static final long ID_MIN = Long.MIN_VALUE >> 3;
+    public static final long ID_MAX = Long.MAX_VALUE >> 3;
+    public static final long ID_MASK = 0x0fffffffffffffffL;
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BytesBuffer.java]: public static final long ID_MASK = 0x7fffffffffffffffL >> 3;

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BytesBuffer.java [hunk 4]
```diff
@@ -355,31 +364,47 @@ public BinaryId parseId() {
...
-            return this.read();
-        } else if (len <= 2) {
-            return this.readShort();
-        } else if (len <= 4) {
-            return this.readInt();
-        } else {
-            assert len == 8 : len;
-            return this.readLong();
+    private long readNumber(int b) {
+        // Parse length from byte 0x0llsnnnn: bits `ll` is the number length
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BytesBuffer.java]: 0b

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BytesBuffer.java [hunk 5]
```diff
@@ -355,31 +364,47 @@ public BinaryId parseId() {
...
+        int length = b >> 5;
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BytesBuffer.java]: check highest bit is 0

**Pull request merged**